components/builder-api/src/server/handlers.rs

```diff
@@ -175,16 +184,28 @@ pub fn job_group_promote(req: &mut Request) -> IronResult<Response> {
         None => return Ok(Response::with(status::BadRequest)),
     };
 
-    let idents = match req.get::<bodyparser::Struct<GroupPromoteReq>>() {
-        Ok(Some(gpr)) => Some(gpr.idents),
-        Ok(None) => None,
-        Err(err) => {
-            debug!("Error decoding json struct: {:?}", err);
-            return Ok(Response::with(status::BadRequest));
-        }
-    };
+    let idents;
+    if promote {
+        idents = match req.get::<bodyparser::Struct<GroupPromoteReq>>() {
+            Ok(Some(gpr)) => Some(gpr.idents),
+            Ok(None) => None,
+            Err(err) => {
+                debug!("Error decoding json struct: {:?}", err);
+                return Ok(Response::with(status::BadRequest));
+            }
+        };
+    } else {
+        idents = match req.get::<bodyparser::Struct<GroupDemoteReq>>() {
+            Ok(Some(gpr)) => Some(gpr.idents),
+            Ok(None) => None,
+            Err(err) => {
+                debug!("Error decoding json struct: {:?}", err);
+                return Ok(Response::with(status::BadRequest));
+            }
+        };
+    }
```

Comment:
It would've been nice to make this more general, but since it's a type, we can't do `req.get::<bodyparser::Struct<T>>()` unless we make `job_group_promote_or_demote` generic over `T` and the trait bounding for that seemed hairy to the point of making the whole thing less readable. I'd appreciate any feedback on whether theres' a cleaner way, though.

Reply:
It seems fine to me as is.  One thing you _can_ do though, if you're interested in a tiny cleanup, is:
```rust
let idents = if promote {
...
} else {
...
};
```